R/extract.R

```diff
@@ -0,0 +1,174 @@
+#'    whether the fitted or original recipe is returned.
+#'
+#' @param x A workflow
+#' @param estimated A logical for whether the original (unfit) recipe or the
+#' fitted recipe should be returned.
+#' @param ... Not currently used.
+#' @details
+#' These functions supersede the `pull_*()` functions.
+#'
+#' Extracting the underlying engine fit can be helpful for describing the
+#'  model (via `print()`, `summary()`, `plot()`, etc.) or for variable
+#'  importance/explainers.
+#'
+#' However, users should not invoke the `predict()` method on an extracted
+#'  model. There may be preprocessing operations that `workflows` has executed on
+#'  the data prior to giving it to the model. Bypassing these can lead to errors
+#'  or silently generating incorrect predictions.
+#'
+#' *Good*:
+#' ```r
+#'    workflow_fit %>% predict(new_data)
+#' ```
+#'
+#' *Bad*:
+#' ```r
+#'    workflow_fit %>% extract_fit_engine()  %>% predict(new_data)
+#'    # or
+#'    workflow_fit %>% extract_fit_parsnip() %>% predict(new_data)
+#' ```
+#' @return
+#' The extracted value from the object, `x`, as described in the description
+#' section.
+#'
+#' @name extract-workflow
+#' @examples
+#' library(parsnip)
+#' library(recipes)
+#' library(magrittr)
+#'
+#' model <- linear_reg() %>%
+#'   set_engine("lm")
+#'
+#' recipe <- recipe(mpg ~ cyl + disp, mtcars) %>%
+#'   step_log(disp)
+#'
+#' base_wf <- workflow() %>%
+#'   add_model(model)
+#'
+#' recipe_wf <- add_recipe(base_wf, recipe)
+#' formula_wf <- add_formula(base_wf, mpg ~ cyl + log(disp))
+#' variable_wf <- add_variables(base_wf, mpg, c(cyl, disp))
+#'
+#' fit_recipe_wf <- fit(recipe_wf, mtcars)
+#' fit_formula_wf <- fit(formula_wf, mtcars)
+#'
+#' # The preprocessor is a recipes, formula, or a list holding the
+#' # tidyselect expressions identifying the outcomes/predictors
+#' extract_preprocessor(recipe_wf)
+#' extract_preprocessor(formula_wf)
+#' extract_preprocessor(variable_wf)
+#'
+#' # The `spec` is the parsnip spec before it has been fit.
+#' # The `fit` is the fit parsnip model.
+#' extract_spec_parsnip(fit_formula_wf)
+#' extract_fit_parsnip(fit_formula_wf)
+#' extract_fit_engine(fit_formula_wf)
+#'
+#' # The mold is returned from `hardhat::mold()`, and contains the
+#' # predictors, outcomes, and information about the preprocessing
+#' # for use on new data at `predict()` time.
+#' extract_mold(fit_recipe_wf)
+#'
+#' # A useful shortcut is to extract the fitted recipe from the workflow
+#' extract_recipe(fit_recipe_wf)
+#'
+#' # That is identical to
+#' identical(
+#'   extract_mold(fit_recipe_wf)$blueprint$recipe,
+#'   extract_recipe(fit_recipe_wf)
+#' )
+NULL
+
+#' @export
+#' @rdname extract-workflow
+extract_spec_parsnip.workflow <- function(x, ...) {
+  if (has_spec(x)) {
+    return(x$fit$actions$model$spec)
+  }
+  abort("The workflow does not have a model spec.")
+}
+
+#' @export
+#' @rdname extract-workflow
+extract_recipe.workflow <- function(x, estimated = TRUE, ...) {
```

Comment:
I feel somewhat strongly that the signature should be:

```suggestion
extract_recipe.workflow <- function(x, ..., estimated = TRUE) {
```

With a call to `ellipsis::check_dots_empty()` inside the method.

I think that if a user put `extract_recipe(wflow, FALSE)` in production code then that would be horrible to come back to in 6 months